test/predefined_dhparams.c

```diff
@@ -0,0 +1,172 @@
+/*
+ * Copyright 1995-2020 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <openssl/evp.h>
+#include <openssl/core_names.h>
+#include <openssl/param_build.h>
+
+#include "predefined_dhparams.h"
+
+#ifndef OPENSSL_NO_DH
+
+static EVP_PKEY *get_dh_from_pg_bn(OSSL_LIB_CTX *libctx, const char *type,
+                                   BIGNUM *p, BIGNUM *g, BIGNUM *q)
+{
+    EVP_PKEY_CTX *pctx = EVP_PKEY_CTX_new_from_name(libctx, type, NULL);
+    OSSL_PARAM_BLD *tmpl = NULL;
+    OSSL_PARAM *params = NULL;
+    EVP_PKEY *dhpkey = NULL;
+
+    if (pctx == NULL || !EVP_PKEY_key_fromdata_init(pctx))
+        goto err;
+
+    if ((tmpl = OSSL_PARAM_BLD_new()) == NULL
+            || !OSSL_PARAM_BLD_push_BN(tmpl, OSSL_PKEY_PARAM_FFC_P, p)
+            || !OSSL_PARAM_BLD_push_BN(tmpl, OSSL_PKEY_PARAM_FFC_G, g)
+            || (q != NULL
+                && !OSSL_PARAM_BLD_push_BN(tmpl, OSSL_PKEY_PARAM_FFC_Q, q)))
+        goto err;
+
+    params = OSSL_PARAM_BLD_to_param(tmpl);
+    if (params == NULL || !EVP_PKEY_fromdata(pctx, &dhpkey, params))
+        goto err;
+
+ err:
+    EVP_PKEY_CTX_free(pctx);
+    OSSL_PARAM_BLD_free_params(params);
+    OSSL_PARAM_BLD_free(tmpl);
+    return dhpkey;
+}
+
+static EVP_PKEY *get_dh_from_pg(OSSL_LIB_CTX *libctx, const char *type,
+                                unsigned char *pdata, size_t plen,
+                                unsigned char *gdata, size_t glen,
+                                unsigned char *qdata, size_t qlen)
+{
+    EVP_PKEY *dhpkey = NULL;
+    BIGNUM *p = NULL, *g = NULL, *q = NULL;
+
+    p = BN_bin2bn(pdata, plen, NULL);
+    g = BN_bin2bn(gdata, glen, NULL);
+    if (p == NULL || g == NULL)
+        goto err;
+    if (qdata != NULL && (q = BN_bin2bn(qdata, qlen, NULL)) == NULL)
+        goto err;
+
+    dhpkey = get_dh_from_pg_bn(libctx, type, p, g, q);
+
+ err:
+    BN_free(p);
+    BN_free(g);
+    BN_free(q);
+    return dhpkey;
+}
+
+EVP_PKEY *get_dh512(OSSL_LIB_CTX *libctx)
```

Comment:
I would have put a disclaimer comment on this..
i.e: This is for testing only.. Use a minimum key size of 2048 for security purposes.

Reply:
Good point, will do.

Reply:
@slontis, this is meanwhile done in 1234aa7e415e1e239eb1c4504578ab59d90763ea.